infra/token-transfer-server/src/lib/lockup.js

```diff
@@ -19,14 +25,50 @@ const { encryptionSecret, clientUrl } = require('../config')
  * Adds a lockup
  * @param userId - user id of the user adding the lockup
  * @param amount - the amount to be locked
+ * @param early - whether this is an early lockup of the next vest
+ * @param data - additional data to be stored with the lockup, e.g. fingerprint
  * @returns {Promise<Lockup>} Lockup object.
  */
-async function addLockup(userId, amount, data = {}) {
-  const user = await hasBalance(userId, amount)
+async function addLockup(userId, amount, early, data = {}) {
+  // Check if user has sufficient balance
+  if (early) {
+    // This is an early lockup for the next vest so call alternate balance
+    // checking function
+    const balance = await getNextVestBalance(userId, amount)
+    if (BigNumber(amount).gt(balance)) {
+      throw new RangeError(
+        `Amount of ${amount} OGN exceeds the ${balance} available for early lockup for user ${userId}`
+      )
+    }
+
+    // Augment lockup data field with additional information about the future
+    // vesting event that this lockup is being created for
+    const user = await User.findOne({
+      where: {
+        id: userId
+      },
+      include: [{ model: Grant }]
+    })
+    data = {
```

Comment:
I wonder if we should also store an "early" boolean flag in the data, to make it easy to query/look at early lockups.

Reply:
I'll probably leave this for now, I don't think we'll be doing manual queries and its a pretty small optimisation, this is actually less code than adding the migration, thanks for suggesting it though